content: draft: Create dependency-track.md (#1432)
slsa-framework/slsa: docs/spec/draft/slsa-dependency-track.md

Reviewer: Oh it looks like there are linter errors.  You can run the linter to fix them.  https://github.com/slsa-framework/slsa/blob/main/CONTRIBUTING.md#markdown-style

Author: We are actively addressing the linter errors. Marking this as resolved.

Reviewer 2: Created https://github.com/slsa-framework/slsa/pull/1446 for the Tracks Page

Reviewer 2: I have a draft  for the Nav bar - i'll submit the PR when we know the URLs and structure

Author: Tracks page has been merged. Will update main.yml immediately after dependency-track is merged

# SLSA Dependency Track v0.1

## Objective

Enable a software producer to measure, control, and reduce risk introduced from third party dependencies.

## Audience

The Dependency Track is primarily aimed at enterprises/organizations, with medium to large-sized organizations benefiting the most from adoption of the dependency track levels. Organizations include enterprises, but also large open source projects that want to manage third party dependency risk.

## Scope

The scope of the SLSA Dependency Track is to properly manage and securely consume build dependencies into the developer workflow to mitigate against supply chain threats targeting the open source ecosystem.

> **Build dependencies:** Software artifacts fetched or otherwise made available to the build environment during the build of the artifact. This includes open and closed source binary dependencies.

Reviewer 2: ```suggestion
```

---

## Definitions

| Primary Term                | Description                                                                                                                                      |
|-----------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------|

> | **Package registry**        | An entity responsible for mapping package names to artifacts within a packaging ecosystem. Most ecosystems support multiple registries, usually a single global registry and multiple private registries. |

Reviewer 2: ```suggestion

|**Build dependencies:**        |Software artifacts fetched or otherwise made available to the build environment during the build of the artifact. This includes open and closed source binary dependencies.|
| **Package registry**        | An entity responsible for mapping package names to artifacts within a packaging ecosystem. Most ecosystems support multiple registries, usually a single global registry and multiple private registries. |
```

## Level 0: Uncontrolled dependency consumption
Dependencies are consumed without any verification, security checks, governance, or controls. This exposes systems to malicious or vulnerable components.

---

## Level 1: Maintain a complete inventory of all build dependencies for a released artifact.

**Outcome:**  
All third party build dependencies (including transitive) are identified such that a software producer is able to automatically match them against known vulnerabilities and assess exposure to supply chain incidents (e.g. compromised upstream open source package). Build dependencies can be identified by:
- Ecosystem, name and version (e.g. native names and versions used in package registries).
- Common Platform Enumeration (CPE)
- Other identifiers that enable joining with known vulnerabilities data sources.

**Intended for:**
Consumers of third party dependencies, including open source, and third party binaries wanting to ingest dependencies from trusted and authenticated reputable package registries.

**Compliance options:**
- Build-time SBOM generated by the build tool, capturing complete direct and transitive dependencies for the artifact being released.

> - For application-level dependencies, dependency pinning can serve as a starting point for many ecosystems as it provides identification for all transitive dependencies (e.g. `package-lock.json`, `pip –require-hashes`, `go.sum`, Gradle lock).

Reviewer 3: ```suggestion
- For language ecosystem dependencies, dependency pinning can serve as a starting point for many ecosystems as it provides identification for all transitive dependencies (e.g. `package-lock.json`, `pip –require-hashes`, `go.sum`, Gradle lock).
```

---

## Level 3: All third-party build dependencies are consumed from sources under the producer’s control.

**Outcome:**  
- The build process consumes all third party build dependencies only from producer-controlled sources, which must address availability risks and third-party dependencies are mirrored from their canonical sources. 
- Gains control of how dependencies enter the supply chain, reducing your attackable surface through standardization of dependency consumption. 

**Intended for:**
Consumers of third party dependencies, including open source, and third party binaries wanting to improve business continuity and control the consumption of dependencies through a standardized consumption method. 

**Compliance options:**
- Use of artifact management solutions to internally mirror third party dependencies in packaged form.
- Source code vendoring, e.g. by mirroring a copy of the upstream code in a local VCS or using ecosystems tools like `go mod vendor` or `cargo vendor`.
- Republishing third-party Docker base image under producer’s project.

**NOTES / Discussion:**
- Build step must enforce use of systems under producer’s control.

> - Compliance with this level enables Level 4.

Reviewer 2: ```suggestion
- Compliance with this level enables the ability to achieve Level 4 compliance.
```

- Leverage dependency security metadata feeds to inform policy decisions on dependencies you consume.

**NOTES / Discussion:**
- This capability builds on Level 3.

**Benefits:**
All of Dependency L3, plus:
- Policy enforcement systems are integrated into the standardized consumption method to check dependencies for issues before the dependency is downloaded.
- Reduces risk of initial compromise from supply chain threats, preventing costs incurred from responding to an incident.

**Threats mitigated at this level:**
- A malicious actor compromises a known good dependency to add malicious functionality in a newly published version of that dependency (e.g. ESLint incident)
- A dependency adds a new transitive dependency that is malicious (e.g. Event-Stream incident)
- Typosquatting
- Dependency Confusion
- Intentional vulnerabilities/backdoors/protestware added to a code base (e.g. Colors v1.4.1 incident, node-ipc incident)

---

## Requirements

| **Requirement** | **Description** | **L1** | **L2** | **L3** | **L4** |
|-----------------|-----------------|--------|--------|--------|--------|

> | Maintain an automated, complete inventory of all dependencies used in development | An organization producing artifacts MUST implement automated tooling that inventories dependencies for every version they release | ✓ | ✓ | ✓ | ✓ |

Reviewer 2: ```suggestion
| Maintain a complete inventory of all dependencies used in development | An organization producing artifacts MUST implement tooling that inventories dependencies for every version they release | ✓ | ✓ | ✓ | ✓ |
```
```suggestion
| Maintain an automated, complete inventory of all dependencies used in development | An organization producing artifacts MUST implement automated tooling that inventories dependencies for every version they release | ✓ | ✓ | ✓ | ✓ |
```

| **Dependency security metadata feed** | A metadata feed that provides security or risk information about each dependency to assist consumers in making judgements about the safety of their dependencies |
| **Package source files**    | A language-specific config file in a source code repo that identifies the package registry feeds or artifact repository manager feeds to consume dependencies from |

---

## Onboarding

When onboarding your development environment to the SLSA Dependency Track, organizations are making claims about how dependencies are managed and consumed from that artifact version forward. This establishes continuity.

No claims are made for prior artifact versions.

## Levels and tracks

| **Track/Level** | **Requirements** | 
|-----------------|------------------|
| Dependency L0 | (none) |
| Dependency L1 | Complete dependency inventory exists |
| Dependency L2 | Known vulnerabilities have been triaged |
| Dependency L3 | Dependencies consumed from sources under producer's control |

> | Dependency L4 | Dependencies meet minimum security requirements |

Reviewer 2: ```suggestion
| Dependency L4 | Proactive defence against upstream attacks. |
```
